op-plasma/damgr.go

```diff
@@ -2,48 +2,416 @@ package plasma
+	ResolveWindow uint64
 }
 
 type DA struct {
 	log     log.Logger
+	cfg     Config
+	metrics Metricer
+
 	storage DAStorage
-}
+	l1      L1Fetcher
+
+	state *State
 
-type Input struct {
-	Data eth.Data
+	// the latest l1 block we synced challenge contract events from
+	origin eth.BlockID
+	// the latest recorded finalized head as per the challenge contract
+	finalizedHead eth.L1BlockRef
+	// the latest recorded finalized head as per the l1 finalization signal
+	l1FinalizedHead eth.L1BlockRef
+	// flag the reset function we are resetting because of an expired challenge
+	resetting int
+
+	finalizedHeadSignalFunc eth.HeadSignalFn
 }
 
 // NewPlasmaDA creates a new PlasmaDA instance with the given log and CLIConfig.
-func NewPlasmaDA(log log.Logger, cfg CLIConfig) *DA {
+func NewPlasmaDA(log log.Logger, cli CLIConfig, cfg Config, l1f L1Fetcher, metrics Metricer) *DA {
+	return NewPlasmaDAWithStorage(log, cfg, cli.NewDAClient(), l1f, metrics)
+}
+
+// NewPlasmaDAWithStorage creates a new PlasmaDA instance with the given log and DAStorage interface.
+func NewPlasmaDAWithStorage(log log.Logger, cfg Config, storage DAStorage, l1f L1Fetcher, metrics Metricer) *DA {
 	return &DA{
 		log:     log,
-		storage: cfg.NewDAClient(),
+		cfg:     cfg,
+		storage: storage,
+		l1:      l1f,
+		metrics: metrics,
+		state:   NewState(log, metrics),
 	}
 }
 
-// NewPlasmaDAWithStorage creates a new PlasmaDA instance with the given log and DAStorage interface.
-func NewPlasmaDAWithStorage(log log.Logger, storage DAStorage) *DA {
+// NewPlasmaWithState creates a plasma storage from initial state used for testing in isolation.
+func NewPlasmaDAWithState(log log.Logger, cfg Config, storage DAStorage, l1f L1Fetcher, metrics Metricer, state *State) *DA {
 	return &DA{
 		log:     log,
+		cfg:     cfg,
 		storage: storage,
+		l1:      l1f,
+		metrics: metrics,
+		state:   state,
+	}
+}
+
+// OnFinalizedHeadSignal sets the callback function to be called when the finalized head is updated.
+// This will signal to the engine queue that will set the proper L2 block as finalized.
+func (d *DA) OnFinalizedHeadSignal(f eth.HeadSignalFn) {
+	d.finalizedHeadSignalFunc = f
+}
+
+// FinalizeL1 advances l1 finalized head in case l1 finalization signal is behind DA finality.
+// If DA challenge params define a shorter finality period, the finalized head will update based on
+// L1 finality.
+func (d *DA) FinalizeL1(ref eth.L1BlockRef) {
+	d.log.Warn("l1 finalized", "ref", ref)
+	if ref.Number > d.l1FinalizedHead.Number {
+		d.l1FinalizedHead = ref
 	}
 }
 
+// LookAhead increments the challenges origin and process the new block if it exists.
+// It is used when the derivation pipeline stalls due to missing data and we need to continue
+// syncing challenge events until the challenge is resolved or expires.
+func (d *DA) LookAhead(ctx context.Context) error {
+	blkRef, err := d.l1.L1BlockRefByNumber(ctx, d.origin.Number+1)
+	if errors.Is(err, ethereum.NotFound) {
+		return io.EOF
```

Comment:
Returning `io.EOF` means that this stage of the pipeline cannot advance any further. Is that a desired behavior to stop looking ahead for this data and instead proceed through the rest of derivation.

I think that this probably breaks the invariant that we process batches in order on L1.